src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs

```diff
@@ -36,42 +36,38 @@ protected internal override void Setup(FeatureConfigurationContext context)
-                var messageTypesHandled = GetMessageTypesHandledByThisEndpoint(handlerRegistry, conventions, settings);
-                return new ApplySubscriptions(messageTypesHandled);
-            });
-        }
 
-        static Type[] GetMessageTypesHandledByThisEndpoint(MessageHandlerRegistry handlerRegistry, Conventions conventions, SubscribeSettings settings)
-        {
-            var messageTypesHandled = handlerRegistry.GetMessageTypes() //get all potential messages
-                .Where(t => !conventions.IsInSystemConventionList(t)) //never auto-subscribe system messages
-                .Where(t => !conventions.IsCommandType(t)) //commands should never be subscribed to
-                .Where(t => conventions.IsEventType(t)) //only events unless the user asked for all messages
-                .Where(t => settings.AutoSubscribeSagas || handlerRegistry.GetHandlersFor(t).Any(handler => !typeof(Saga).IsAssignableFrom(handler.HandlerType))) //get messages with other handlers than sagas if needed
-                .Except(settings.ExcludedTypes)
-                .ToArray();
-
-            return messageTypesHandled;
+                return new ApplySubscriptions(handlerRegistry, conventions, settings);
+            });
         }
 
         class ApplySubscriptions : FeatureStartupTask
         {
-            public ApplySubscriptions(Type[] messagesHandledByThisEndpoint)
+            public ApplySubscriptions(MessageHandlerRegistry messageHandlerRegistry, Conventions conventions, SubscribeSettings subscribeSettings)
             {
-                this.messagesHandledByThisEndpoint = messagesHandledByThisEndpoint;
+                this.messageHandlerRegistry = messageHandlerRegistry;
+                this.conventions = conventions;
+                this.subscribeSettings = subscribeSettings;
             }
 
             protected override async Task OnStart(IMessageSession session, CancellationToken cancellationToken = default)
             {
+                var eventsToSubscribe = GetHandledEventTypes(messageHandlerRegistry, conventions, subscribeSettings);
+
+                if (eventsToSubscribe.Length == 0)
+                {
+                    return;
+                }
+
                 try
                 {
                     var messageSession = session as MessageSession;
-                    await messageSession.SubscribeAll(messagesHandledByThisEndpoint, new SubscribeOptions(), cancellationToken)
+                    await messageSession.SubscribeAll(eventsToSubscribe, new SubscribeOptions(), cancellationToken)
                         .ConfigureAwait(false);
                     if (Logger.IsDebugEnabled)
                     {
                         Logger.DebugFormat("Auto subscribed to events {0}",
-                            string.Join<Type>(",", messagesHandledByThisEndpoint));
+                            string.Join<Type>(",", eventsToSubscribe));
```

Comment:
We used to log this even if there were no events. Would it make sense to include a message above (line 58ish) to say that autosubscribe found no messages to subscribe to?

Reply:
> We used to log this even if there were no events. Would it make sense to include a message above (line 58ish) to say that autosubscribe found no messages to subscribe to?

yeah that's a good idea, would make more sense than logging "autosubscribed to events" when the list is empty